### PR TITLE
Fix multi gpu map example

### PR DESCRIPTION
**Pull request by lhoestq**
- use `orch.cuda.set_device` instead of `CUDA_VISIBLE_DEVICES `
- add `if __name__ == "__main__"`

fix https://github.com/huggingface/datasets/issues/6186

### CODE REVIEW

File: docs/source/process.mdx [hunk 1]
```diff
@@ -340,21 +340,29 @@ You can also use [`~Dataset.map`] with indices if you set `with_indices=True`. T
...
 >>> from multiprocess import set_start_method
 >>> import torch
 >>> import os
 >>>
->>> set_start_method("spawn")
->>>
 >>> def gpu_computation(example, rank):
->>>     os.environ["CUDA_VISIBLE_DEVICES"] = str(rank % torch.cuda.device_count())
+>>>     torch.cuda.set_device(rank % torch.cuda.device_count())
 >>>     # Your big GPU call goes here
```
Comment (NielsRogge) [inline docs/source/process.mdx:360]: I still would like to see a concrete example here instead of "your big GPU call goes here", cause I tried using an NLLB model with 2 GPUs to translate sentences of the [datacomp dataset](https://huggingface.co/datasets/mlfoundations/datacomp_small) in parallel and it was unclear for me how I had to do it. Should I use nn.DataParallel? Should I use .to("cuda:0") and .to("cuda:1")?

Comment (NielsRogge) [reply docs/source/process.mdx:360]: I remember that the rank was always set to 0, so all work was done on the first GPU

Comment (lhoestq) [reply docs/source/process.mdx:360]: > I remember that the rank was always set to 0, so all work was done on the first GPU

This happens only if you set num_proc=1, but for multiprocessing you get multiple ranks (one per process)

Comment (lhoestq) [reply docs/source/process.mdx]: I added `model` in the example - no need to use nn.DataParallel. You just need to send the model to every GPU.

Feel free to test that the code works as expected for you !

File: docs/source/process.mdx [hunk 2]
```diff
@@ -340,21 +340,36 @@ You can also use [`~Dataset.map`] with indices if you set `with_indices=True`. T
...
 The [`~Dataset.map`] also works with the rank of the process if you set `with_rank=True`. This is analogous to the `with_indices` parameter. The `with_rank` parameter in the mapped function goes after the `index` one if it is already present. 
 
 ```py
 >>> from multiprocess import set_start_method
 >>> import torch
 >>> import os
 >>>
->>> set_start_method("spawn")
+>>> for i in range(torch.cuda.device_count()):  # send model to every GPU
+...     model.to(torch.cuda.device(i))
```
Comment (NielsRogge) [inline docs/source/process.mdx:359]: This gives me the following error:
```
Device: <torch.cuda.device object at 0x7f49aca62920>
Traceback (most recent call last):
  File "/home/niels/python_projects/datacomp/datasets_multi_gpu.py", line 14, in <module>
    model.to(torch.cuda.device(i))
  File "/home/niels/anaconda3/envs/datacomp/lib/python3.10/site-packages/torch/nn/modules/module.py", line 968, in to
    device, dtype, non_blocking, convert_to_format = torch._C._nn._parse_to(*args, **kwargs)
TypeError: to() received an invalid combination of arguments - got (device), but expected one of:
 * (torch.device device, torch.dtype dtype, bool non_blocking, bool copy, *, torch.memory_format memory_format)
 * (torch.dtype dtype, bool non_blocking, bool copy, *, torch.memory_format memory_format)
 * (Tensor tensor, bool non_blocking, bool copy, *, torch.memory_format memory_format)
```

Comment (lhoestq) [reply docs/source/process.mdx]: fixed it, thanks

**Comment by lhoestq** [comment]
Merging this one, but lmk if you have more comments for subsequent improvements @NielsRogge

**Comment by alex2awesome** [comment]
This is a little hard to follow — where is the documentation currently? I am trying to follow from here:

```>>> import os
>>>
>>> for i in range(torch.cuda.device_count()):  # send model to every GPU
...     model.to(f"cuda:{i}")
>>>
>>> def gpu_computation(example, rank):
...     torch.cuda.set_device(f"cuda:{rank}")  # use one GPU
...     inputs = tokenizer(texts, truncation=True, return_tensors="pt").to(f"cuda:{rank}")
...     outputs = model(**inputs)
...      .... 
```

but I'm getting device errors (data is on device 3, but it thinks model is on device 0, despite setting `torch.cuda.set_device`

Is this correct? What version of Torch are you using for this?

**Comment by alex2awesome** [comment]
Anyway, this didn't work for me:

`torch.cuda.set_device(f"cuda:{rank}")  # use one GPU`

but this did:

`model.to(f"cuda:{rank}")`

**Comment by NielsRogge** [comment]
Yeah for me this issue isn't resolved yet, we need a better code example

**Comment by lhoestq** [comment]
Hi @alex2awesome, could you open a PR with your suggestion to improve this code snippet ?

**Comment by alex2awesome** [comment]
i'm happy to when i get it fully working, but i feel like there are some fundamentals that I'm not fully understanding...

I've set it up twice now, for 2 GPU-processing pipelines. 

In one pipelines, my memory usage is fine, it delivers me a huge speedup, and everything is great. In the second pipeline, I keep getting OOM errors when `num_proc > 1` that I don't get when `num_proc=1`. 

There is a discussion here: https://github.com/pytorch/pytorch/issues/44156 about CUDA memory leaks in multiprocessing setups, and I haven't had the time to fully read the source code to `datasets.map` to understand whether the situations are parallel.

I haven't fully tested out enough to see what the issue is but if I feel comfortable over the next several days to generate a slimmed-down example that I know will generalize to the real-world cases I'm working with now, then I will contribute it.

**Comment by alex2awesome** [comment]
@lhoestq do you know how `datasets` does multiprocessing? Do we use:
https://pytorch.org/docs/stable/multiprocessing.html#module-torch.multiprocessing?

If so, there are lots of points around memory usage, here:
https://pytorch.org/docs/stable/notes/multiprocessing.html

**Comment by alex2awesome** [comment]
After some more research and playing around, I can't pinpoint the source of my CUDA memory leak and I'm not really an expert on multiprocessing in general, but my gut is that the current set-up isn't ideal for multiprocessing. 

Are either of you more experienced in this?

**Comment by lhoestq** [comment]
Not sure about your GPU's OOM :/

Still, I opened a PR with your suggestion here: https://github.com/huggingface/datasets/pull/6550

**Comment by kopyl** [comment]
I still get only 0 rank...

Here is my code: https://pastebin.com/c6du8jaM

from this ^ i just improt one function:


```
from test import map_train
from multiprocess import set_start_method


set_start_method("spawn")
map_train()
```

And here is the traceback:
https://pastebin.com/YijspwQK

**Comment by kopyl** [comment]
Also this code from your docs is not valid (source: https://huggingface.co/docs/datasets/main/en/process#multiprocessing):
```
for i in range(torch.cuda.device_count()):
    model.to(f"cuda:{i}")
```


This for me sends the model only to the second GPU
```
vae = AutoencoderKL.from_pretrained(
    pretrained_model_name_or_path, subfolder="vae"
)
vae.to("cuda:0")
vae.to("cuda:1")
```

**Comment by kopyl** [comment]
Could you please provide a working example of multi-GPU mapping?

Not just an example in docs, but a real working example starting from all imports and datasets.

**Comment by kopyl** [comment]
@alex2awesome the same issue with CUDA OOM. It should not be happening, since it should 2 different GPUs be handling different loads. But in fact something wrong is happening.

**Comment by lhoestq** [comment]
I haven't experimented much with the multi-GPU code documentation (see discussion at #6550)

Though if you have some fixes/improvements ideas feel free to open a PR !

**Comment by kopyl** [comment]
@lhoestq the mapping does not start at all in this case:
image

**Comment by kopyl** [comment]
@lhoestq with this code: https://pastebin.com/muDm78kp
i now getting this error:

```
Map (num_proc=2):   1%|       | 26288/3043663 [06:11<11:51:08, 70.72 examples/s]
Traceback (most recent call last):
  File "/workspace/compute.py", line 229, in <module>
    map_train()
  File "/workspace/compute.py", line 224, in map_train
    return train_dataset.map(compute_embeddings_fn, batched=True, batch_size=16, with_rank=True, num_proc=2, keep_in_memory=True)
  File "/usr/local/lib/python3.10/dist-packages/datasets/arrow_dataset.py", line 593, in wrapper
    out: Union["Dataset", "DatasetDict"] = func(self, *args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/datasets/arrow_dataset.py", line 558, in wrapper
    out: Union["Dataset", "DatasetDict"] = func(self, *args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/datasets/arrow_dataset.py", line 3193, in map
    for rank, done, content in iflatmap_unordered(
  File "/usr/local/lib/python3.10/dist-packages/datasets/utils/py_utils.py", line 658, in iflatmap_unordered
    raise RuntimeError(
RuntimeError: One of the subprocesses has abruptly died during map operation.To debug the error, disable multiprocessing.
```

Also when trying to download my dataset there were no issues from one machine, but from another:
```
SSLError: (MaxRetryError("HTTPSConnectionPool(host='huggingface.co', port=443): Max retries exceeded with url: /api/datasets/kopyl/3M_icons_monochrome_only_no_captioning/revision/753dca4be462dad7022f34cc273555ab6deb5832 (Caused by SSLError(SSLEOFError(8, '[SSL: UNEXPECTED_EOF_WHILE_READING] EOF occurred in violation of protocol (_ssl.c:1007)')))"), '(Request ID: 7d0881f3-1b93-4d73-bcb6-52e816d84529)')
```

Can't download my dataset at all...

**Comment by lhoestq** [comment]
Hmm this is not good, do you know a way to make it work ?

Basically `map` creates two subprocesses and runs the function in the subprocesses. Since each function has a parameter `rank` it should be possible to choose which GPU to use

**Comment by forrestbao** [comment]
I can confirm that PR #6550 works. All GPUs are at full throttle. You have to manually move the model to all GPUs. 

> I haven't experimented much with the multi-GPU code documentation.
> 
> Can you try using the code example at #6550 instead ? That would be super helpful if you could confirm that it works on your side
> 
> Though if you have some fixes/improvements ideas feel free to open a PR !

**Comment by forrestbao** [comment]
I wrote a [blog post](https://forrestbao.github.io/2024/01/30/datasets_map_with_rank_multiple_GPUs.html) with a complete example by compiling information from several PRs and issues here. Hope it can help. Let me know how it works. 

> Could you please provide a working example of multi-GPU mapping?
> 
> Not just an example in docs, but a real working example starting from all imports loading datasets and models.
